inference-engine/src/inference_engine/src/cpp_interfaces/interface/ie_iinfer_request_internal.cpp

```diff
@@ -274,14 +270,22 @@ void IInferRequestInternal::checkBlob(const Blob::Ptr& blob,
             if (foundOutputPair == std::end(_networkOutputs)) {
                 IE_THROW(NotFound) << "Failed to find output with name: \'" << name << "\'";
             }
-            dims = foundOutputPair->second->getTensorDesc().getDims();
+            isDynamic = foundOutputPair->second->getPartialShape().is_dynamic();
+            ngraph::PartialShape blobPartialShape(blob->getTensorDesc().getDims());
+            if (foundOutputPair->second->getPartialShape().compatible(blobPartialShape)) {
+                dims = blob->getTensorDesc().getDims();
+            } else {
+                // TODO: it is strange to request tensor desc from data when the shapes are not compatible, probably we
+                // need to immediately throw here
+                dims = foundOutputPair->second->getTensorDesc().getDims();
+            }
             refSize = foundOutputPair->second->getTensorDesc().getLayout() != SCALAR ? details::product(dims) : 1;
         }
     } else {
         refSize = details::product(refDims);
     }
 
-    if (refSize != blob->size()) {
+    if (!isDynamic && refSize != blob->size()) {
         IE_THROW() << strNotMatched + ": got " << blob->size() << " expecting " << refSize;
     }
     const bool remoteBlobPassed = blob->is<RemoteBlob>();
```

Comment:
comment for `if (!remoteBlobPassed && blob->buffer() == nullptr)` check

we need to call `checkBlob` inside `GetBlob` method, but before customer has not called `setShape` or `SetBlob`, blob will not be allocated and exception will be thrown
So shouldn't we add some check to avoid this?

Reply:
discussed offline